### PR TITLE
Add support for building Portable Linux binaries/packages in pipeline

### PR DESCRIPTION
**Pull request by wtgodbe**
Contingent on https://github.com/dotnet/corefx/pull/14632

### CODE REVIEW

**Comment by wtgodbe** [comment]
CC @gkhanna79 @chcosta @janvorli

File: buildpipeline/pipeline.json [hunk 1]
```diff
@@ -335,6 +348,19 @@
             "Type": "build/product/",
             "ConfigurationGroup": "Debug"
           }
+        }, 
+        {
+          "Name": "DotNet-CoreFx-Trusted-Linux-Native",
+          "Parameters": {
+            "DockerTag": "ubuntu1404_prereqs_v3",
+            "portableLinux": "-portableLinux"
```
Comment (janvorli) [inline buildpipeline/pipeline.json]: Do I understand it correctly that we perform the portable linux build on ubuntu 14.04? If that's the case, we need to change it to use CentOS 7.1 instead. The point is to build on Linux distro with the oldest libraries, especially libicu.

Comment (wtgodbe) [inline buildpipeline/pipeline.json]: That's correct. I can make the switch - I'm guessing I should do the same for CoreCLR?

Comment (janvorli) [inline buildpipeline/pipeline.json]: Right, that should be the same for CoreCLR.

**Comment by gkhanna79** [comment]
DockerTags are at https://hub.docker.com/r/chcosta/dotnetcore/tags/ - CentOS ones are listed at the very end.

**Comment by wtgodbe** [comment]
@chcosta are the CentOS Docker images from https://hub.docker.com/r/chcosta/dotnetcore/tags/ usable? Looks like they haven't been updated in the last year

File: buildpipeline/DotNet-CoreFx-Trusted-Linux-Native.json [hunk 2]
```diff
@@ -263,7 +263,7 @@
       },
       "inputs": {
         "filename": "docker",
-        "arguments": "run -w=\"$(GitDirectory)\" --name $(DockerContainerName) $(DockerContainerName) $(GitDirectory)/build-managed.sh -OfficialBuildId=$(OfficialBuildId) -- /t:GenerateVersionSourceFile /p:GenerateVersionSourceFile=true",
+        "arguments": "run -w=\"$(GitDirectory)\" --name $(DockerContainerName) $(DockerContainerName) $(GitDirectory)/build-managed.sh -OfficialBuildId=$(OfficialBuildId) $(portableLinux) -- /t:GenerateVersionSourceFile /p:GenerateVersionSourceFile=true",
```
Comment (gkhanna79) [inline buildpipeline/DotNet-CoreFx-Trusted-Linux-Native.json]: I don't think this change is required.

**Comment by wtgodbe** [comment]
@gkhanna79 just updated this to use the CentOS docker image, and to pass '-portableLinux' to build-native and build-packages

**Comment by gkhanna79** [comment]
@ellismg PTAL as well.

**Comment by gkhanna79** [comment]
@wtgodbe Can you please update CoreCLR definition to use CentOS docker tag and queue off a build?

File: buildpipeline/pipeline.json [hunk 3]
```diff
@@ -134,6 +134,19 @@
...
           }
+        },
+        {
+          "Name": "DotNet-CoreFx-Trusted-Linux-Native",
+          "Parameters": {
+            "DockerTag": "centos711503_prereqs_2",
+            "portableLinux": "-portableLinux"
+          },
+          "ReportingParameters": {
+            "OperatingSystem": "Ubuntu 14.04",
```
Comment (ellismg) [inline buildpipeline/pipeline.json]: This doesn't seem right...

File: buildpipeline/pipeline.json [hunk 4]
```diff
@@ -335,6 +348,19 @@
...
           }
+        }, 
+        {
+          "Name": "DotNet-CoreFx-Trusted-Linux-Native",
+          "Parameters": {
+            "DockerTag": "centos711503_prereqs_2",
+            "portableLinux": "-portableLinux"
+          },
+          "ReportingParameters": {
+            "OperatingSystem": "Ubuntu 14.04",
```
Comment (ellismg) [inline buildpipeline/pipeline.json]: Similar problem here.

**Comment by ellismg** [comment]
FWIW, we stopped using the CentOS images in favor of the RHEL ones a while back (since we don't do an explicit build for CentOS, just re-use the RHEL binaries).

File: buildpipeline/pipeline.json (hunk 4 above)
Comment (wtgodbe) [inline buildpipeline/pipeline.json]: Righto, my bad. fixed.

**Comment by wtgodbe** [comment]
@janvorli in light of Matt's comment above, would it be fine to build portable binaries on RHEL? Or should we definitely be using CentOS?

**Comment by janvorli** [comment]
@ellismg what is the version of the libicu installed on the RHEL?

**Comment by ellismg** [comment]
@janvorli It is 50.  Should be the same as CentOS?

```
matell@matell-ub:~/git/core-setup$ docker run -it chcosta/dotnetcore:rhel7_prereqs_2
[root@e23b9e711b56 /]# rpm -qa | grep icu
libicu-devel-50.1.2-15.el7.x86_64
libicu-50.1.2-15.el7.x86_64
[root@e23b9e711b56 /]#
```

**Comment by chcosta** [comment]
From a PipeBuild perspective, LGTM

**Comment by gkhanna79** [comment]
@ellismg Can you confirm what is the version in CentOS image as well?

**Comment by gkhanna79** [comment]
@ellismg The key is here to build on an OS that has the lowest version of libicu. If CentOS 7.1 carries the lowest version, is the same guaranteed for RHEL as well? If so, which RHEL is it? 7.1?

If RHEL cannot guarantee that, are there concerns/known issues in building on CentOS?

**Comment by ellismg** [comment]
>@ellismg The key is here to build on an OS that has the lowest version of libicu. If CentOS 7.1 carries the lowest version, is the same guaranteed for RHEL as well? If so, which RHEL is it? 7.1?
>
>If RHEL cannot guarantee that, are there concerns/known issues in building on CentOS?

RHEL and CentOS carry the same versions of packages (CentOS is a rebuild of RHEL, using the RHEL sources without Red Hat trademarks).  They carry the same version of libicu (50).  That's the lowest version of the platforms we support.  Any changes RHEL makes are going to be consumed by CentOS.

The only concern I have around building on CentOS is it's yet another platform we'd need to support for our official builds.  We'll need to continue to maintain these images in the future.

**Comment by janvorli** [comment]
@ellismg, @gkhanna79 then RHEL is good for our purposes.

**Comment by gkhanna79** [comment]
Excellent! Thanks for the details @ellismg.

@wtgodbe Can you please make the relevant changes to CoreCLR build definition (and queue a build) and also to this PR?

**Comment by wtgodbe** [comment]
Just updated this PR & Merged the change for CoreCLR - here's the queued build https://devdiv.visualstudio.com/DefaultCollection/DevDiv/_build/explorer?_a=summary&buildId=481789

**Comment by wtgodbe** [comment]
https://devdiv.visualstudio.com/DefaultCollection/DevDiv/_build/explorer?_a=summary&buildId=481913 here's the pipebuild for CoreFX

**Pull request merged**